feat: Allow specific indexers in EntityIndexerRegistry

NielDuysters opened the pull request:
### 1. Why is this change necessary?
When you want to exclude indexers to execute on the update of an Entity, you currently have to exclude all the indexers manually.
```php
$context->addExtension(EntityIndexerRegistry::EXTENSION_INDEXER_SKIP,
    new ArrayEntity(
        [
            // All except STOCK_UPDATER.
            ProductIndexer::INHERITANCE_UPDATER,
            ProductIndexer::VARIANT_LISTING_UPDATER,
            ProductIndexer::CHILD_COUNT_UPDATER,
            ProductIndexer::MANY_TO_MANY_ID_FIELD_UPDATER,
            ProductIndexer::CATEGORY_DENORMALIZER_UPDATER,
            ProductIndexer::CHEAPEST_PRICE_UPDATER,
            ProductIndexer::RATING_AVERAGE_UPDATER,
            ProductIndexer::STREAM_UPDATER,
            ProductIndexer::SEARCH_KEYWORD_UPDATER,
        ]
    ),
);

$this->productRepository->update([
    [
        'id' => $productId,
        'stock' => $newStock,
    ]
], $context);
```

### 2. What does this change do, exactly?
This change allows you to whitelist only the indexers you want to execute.
```php
$context->addExtension(EntityIndexerRegistry::EXTENSION_INDEXER_ONLY,
    new ArrayEntity(
        [
            // Only execute STOCK_UPDATER.
            ProductIndexer::STOCK_UPDATER
        ]
    ),
);
        
$this->productRepository->update([
    [
        'id' => $productId,
        'stock' => $newStock,
    ]
], $context);
```

### 3. Describe each step to reproduce the issue or behaviour.
I made a simple subscriber to test this code. Add an item to the cart to trigger.
<details>
<summary>View code</summary>

```php
<?php declare(strict_types=1);

namespace DemoWebshop\Subscriber;

use Psr\Log\LoggerInterface;
use Shopware\Core\Checkout\Cart\Event\AfterLineItemAddedEvent;
use Shopware\Core\Content\Product\DataAbstractionLayer\ProductIndexer;
use Shopware\Core\Framework\DataAbstractionLayer\EntityRepository;
use Shopware\Core\Framework\DataAbstractionLayer\Indexing\EntityIndexerRegistry;
use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
use Shopware\Core\Framework\Struct\ArrayStruct;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;

class MySubscriber implements EventSubscriberInterface
{
    public function __construct(
        private readonly EntityRepository $productRepository,
        private readonly LoggerInterface $logger
    )
    {
    }

    public static function getSubscribedEvents(): array
    {
        // Return the events to listen to as array like this:  <event to listen to> => <method to execute>
        return [
            AfterLineItemAddedEvent::class => 'onAfterLineItemAdded',
        ];
    }

    public function onAfterLineItemAdded(AfterLineItemAddedEvent $event): void
    {
        $context = $event->getContext();
        $cart = $event->getCart();
        $newStock = 30;

        $context->addExtension(EntityIndexerRegistry::EXTENSION_INDEXER_SKIP,
            new ArrayStruct(
                [
                    ProductIndexer::STOCK_UPDATER
                ]
            ),
        );

        foreach ($cart->getLineItems() as $cartItem) {
            try {
                /** @var Product $product */
                $product = $this->productRepository->search(
                    (new Criteria([$cartItem->getId()]))->addAssociation('stock'),
                    $context
                )->getEntities()->first();

                $this->productRepository->update([
                    [
                        'id' => $cartItem->getId(),
                        'stock' => $newStock,
                    ]
                ], $context);
            } catch (\Exception $exception) {
                $this->logger->error($exception->getMessage());
            }
        }
    }
}
```
</details>

### 4. Please link to the relevant issues (if any).
N/A

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfill them.

NielDuysters committed:
entityindexerregistry: Add EXTENSION_INDEXER_ONLY

Allow to index specific indexes instead of blacklisting the ones needed to be excluded.

NielDuysters committed:
entityindexerregistry: Update test

NielDuysters committed:
entityindexerregistry: Changelog

wannevancamp commented on src/Core/Framework/DataAbstractionLayer/Indexing/EntityIndexerRegistry.php:
```diff
@@ -145,7 +146,13 @@ public function refresh(EntityWrittenContainerEvent $event): void
             $message->isFullIndexing = false;
-            self::addSkips($message, $context);
+
+            if ($context->hasExtension(self::EXTENSION_INDEXER_SKIP)) {
```
De `$context->hasExtension(self::EXTENSION_INDEXER_SKIP` check wordt dubbel uitgevoerd. Zowel hier als in de addSkips function. Hetzelfde voor de `$context->hasExtension(self::EXTENSION_INDEXER_ONLY` check.

NielDuysters replied:
Klopt, die if op lijn `EntityIndexerRegistry.php:150` is nodig om afhankelijk van de extension ofwel `addSkips` oftewel `addOnlies` uit te voeren.

De if-check in `addSkips` kan ik denk ik beter niet verwijderen omdat die functie ook nog andere usages heeft:
<img width="1175" alt="afbeelding" src="https://github.com/user-attachments/assets/51c085a3-d34c-4030-9537-a0d1bdb94695" />

Wat ik wel kan doen, is de if statements rond deze calls gewoon verwijderen aangezien de checks inderdaad toch in de functies zelf uitgevoerd worden:
<img width="743" alt="afbeelding" src="https://github.com/user-attachments/assets/5978869f-1aad-4989-8e1b-2d9568aefa4e" />

Weet wel niet of dat per se leesbaarder is, ik laat die keuze aan u over. Maar zal het alvast doorvoeren.

**Edit:** Is eigenlijk prima leesbaar zo zonder die obsolete ifs. Idd zoals je zei niet logisch extra if statements toe te voegen terwijl de checks ook in de `addSkips` en `addOnlies` zitten. De extra ifs die ik eerst had maakte het juist verwarrender. Thanks!

**Commit:** https://github.com/NielDuysters/shopware/pull/1/commits/df939f6b899de05c77ef7c465c5f0b8120a74fb0

NielDuysters replied:
Klopt, die if op lijn `EntityIndexerRegistry.php:150` is nodig om afhankelijk van de extension ofwel `addSkips` oftewel `addOnlies` uit te voeren.

De if-check in `addSkips` kan ik denk ik beter niet verwijderen omdat die functie ook nog andere usages heeft:
<img width="1175" alt="afbeelding" src="https://github.com/user-attachments/assets/51c085a3-d34c-4030-9537-a0d1bdb94695" />

Wat ik wel kan doen, is de if statements rond deze calls gewoon verwijderen aangezien de checks inderdaad toch in de functies zelf uitgevoerd worden:
<img width="743" alt="afbeelding" src="https://github.com/user-attachments/assets/5978869f-1aad-4989-8e1b-2d9568aefa4e" />

Weet wel niet of dat per se leesbaarder is, ik laat die keuze aan u over. Maar zal het alvast doorvoeren.

NielDuysters replied:
Klopt, die if op lijn `EntityIndexerRegistry.php:150` is nodig om afhankelijk van de extension ofwel `addSkips` oftewel `addOnlies` uit te voeren.

De if-check in `addSkips` kan ik denk ik beter niet verwijderen omdat die functie ook nog andere usages heeft:
<img width="1175" alt="afbeelding" src="https://github.com/user-attachments/assets/51c085a3-d34c-4030-9537-a0d1bdb94695" />

Wat ik wel kan doen, is de if statements rond deze calls gewoon verwijderen aangezien de checks inderdaad toch in de functies zelf uitgevoerd worden:
<img width="743" alt="afbeelding" src="https://github.com/user-attachments/assets/5978869f-1aad-4989-8e1b-2d9568aefa4e" />

Weet wel niet of dat per se leesbaarder is, ik laat die keuze aan u over. Maar zal het alvast doorvoeren.

wannevancamp commented on src/Core/Framework/DataAbstractionLayer/Indexing/EntityIndexerRegistry.php:
```diff
@@ -166,6 +173,23 @@ public static function addSkips(EntityIndexingMessage $message, Context $context
+        if (!$onlies instanceof ArrayEntity) {
+            return;
+        }
+        $skip = array_diff($options, $onlies->all());
```
Kunnen we deze logica niet op een lager niveau in de `EntityIndexingMessage` implementeren?

Ik zie dat een `getOnly` functie al beschikbaar is in de `FullEntityIndexerMessage`. Zie `src/Core/Framework/DataAbstractionLayer/Indexing/EntityIndexerRegistry.php:50`. Zelf geen idee wat het verschil is tussen deze 2 type messages maar mss wel interessant om te bekijken?

NielDuysters replied:
Ja awel die `onlies` in `FullEntityIndexerMessage` was ik ook al tegengekomen, en was een beetje verwarrend want het leek in het begin dus al of de feature om enkel specifieke indexers mee te geven al bestond.

Maar `FullEntityIndexerMessage` is niet 100% van toepassing op onze use case denk ik. Wanneer wij via een sw repo een PostUpdate call uitvoeren wordt de `EntityIndexingMessage` gebruikt, dezelfde message waar deze MR betrekking op heeft, niet de `FullEntityIndexerMessage`. Omdat `EntityIndexingMessage` dus gelinkt is aan een PostUpdate bevat deze message ook data zoals de IDs van de geupdate object en de geschreven properties etc.

De `FullEntityIndexerMessage` is volgens mij de message die gebruikt wordt wanneer we de indexes system-wide refreshen via het admin-paneel of cli.

De cli en admin-paneel bevatten ook die opties om zowel skips en onlies mee te geven:
<img width="694" alt="afbeelding" src="https://github.com/user-attachments/assets/c9421eb3-10ff-4008-9f08-3b951c719a33" />
<img width="971" alt="afbeelding" src="https://github.com/user-attachments/assets/d0b0d8ef-694c-4f1a-ac1f-84e514d66e0e" />

Daarom dat `FullEntityIndexerMessage` ook de properties `skip` en `only` heeft, mapt volgens mij op het argument in dat cli commando.
<img width="971" alt="afbeelding" src="https://github.com/user-attachments/assets/c44f7f41-2851-42cb-86e5-6cbfc9e155b1" />

NielDuysters replied:
Maar mijn oorspronkelijke reflex was - net zoals jij aangaf - om ook gewoon in `EntityIndexingMessage` een property `only` toe te voegen idem als in `FullEntityIndexerMessage`.

Het issue is dan dat de complexiteit van deze PR exponentieel omhoog zou gaan omdat we dan de interne API van van heel het entity indexing mechanisme wel serieus aan het aanpassen zijn, wat ook meer error prone is.

De huidige implementatie voegt een functionaliteit toe aan entity indexing door gebruik te maken van de reeds bestaande API zonder hier teveel veranderingen in aan te brengen.

Beetje soort van het SOLID principe **Open/Closed Principle – "open for extension, closed for modification"** . We breiden de API uit zonder ingrijpende wijzigingen te maken en de PR beheerst te houden.

Wat denk jij?

wannevancamp commented on src/Core/Framework/DataAbstractionLayer/Indexing/EntityIndexerRegistry.php:
```diff
@@ -166,6 +173,23 @@ public static function addSkips(EntityIndexingMessage $message, Context $context
+        if (!$context->hasExtension(self::EXTENSION_INDEXER_ONLY)) {
+            return;
+        }
+        $onlies = $context->getExtension(self::EXTENSION_INDEXER_ONLY);
```
Ondanks dat dit een array is wordt in de core de naamgeving `only` gebruikt ipv `onlies`.

NielDuysters replied:
**Commit:** https://github.com/NielDuysters/shopware/pull/1/commits/a26bcc7b9b6b97a38463d968e3e41a20c9c252f3

wannevancamp reviewed (changes requested):
Mooie PR @NielDuysters !

wannevancamp commented:
@NielDuysters , heb je dit nog verder kunnen bekijken?

NielDuysters committed:
entityindexerregistry: Small refactor - Remove obsolete if statements

These if-statements are obsolete because the checks for the extension is done in the body of the called methods as well.

We can consider reverting this commit and keep the if-statements for readability.

NielDuysters committed:
entityindexerregistry: Small refactor - Variable naming convention

NielDuysters commented:
@wannevancamp MR updated.

Bleef even liggen want was ook nog druk met contributen aan andere open-source projecten. Heb voor deze maand bv. een grote contribution gedaan aan [Arti](https://gitlab.torproject.org/tpo/core/arti). Is een project van The Tor Project om de codebase van het Tor netwerk momenteel geschreven in C te herschrijven in Rust.

Ik heb een library geschreven die bridged tussen een populaire http client (ureq) en Arti/Tor. Wat wel een uitdaging was aangezien die http client een blocking I/O heeft en Arti runt in een async non-blocking I/O.

Maar wel leerrijk, en ik wordt normaal ook vermeld in de release-blog volgende week ofzo.

https://gitlab.torproject.org/tpo/core/arti/-/tree/main/crates/arti-ureq

Cool eh! :)
